v3-roadmap-and-design.md

```diff
@@ -0,0 +1,425 @@
+| `__contains__` | `contains` | `__contains__` | `__contains__` |
+| `__getitem__` | `getitem` | `__getitem__` | `__getitem__` |
+| `__enter__` | N/A | N/A | `__enter__` |
+| `__exit__` | N/A | N/A | `__exit__` |
+| `group_keys` | `group_keys` | `group_keys` | N/A |
+| `groups` | `groups` | `groups` | N/A |
+| `array_keys` | `array_key` | `array_keys` | N/A |
+| `arrays` | `arrays`* | `arrays` | N/A |
+| `visit` | ? | ? | `visit` |
+| `visitkeys` | ? | ? | ? |
+| `visitvalues` | ? | ? | ? |
+| `visititems` | ? | ? | `visititems` |
+| `tree` | `tree` | `tree` | `Both` |
+| `create_group` | `create_group` | `create_group` | `create_group` |
+| `require_group` | N/A | N/A | `require_group` |
+| `create_groups` | ? | ? | N/A |
+| `require_groups` | ? | ? | ? |
+| `create_dataset` | N/A | N/A | `create_dataset` |
+| `require_dataset` | N/A | N/A | `require_dataset` |
+| `create` | `create_array` | `create_array` | N/A |
+| `empty` | `empty` | `empty` | N/A |
+| `zeros` | `zeros` | `zeros` | N/A |
+| `ones` | `ones` | `ones` | N/A |
+| `full` | `full` | `full` | N/A |
+| `array` | `create_array` | `create_array` | N/A |
+| `empty_like` | `empty_like` | `empty_like` | N/A |
+| `zeros_like` | `zeros_like` | `zeros_like` | N/A |
+| `ones_like` | `ones_like` | `ones_like` | N/A |
+| `full_like` | `full_like` | `full_like` | N/A |
+| `move` | `move` | `move` | `move` |
+
+**`zarr.h5compat.Group`**
+
+Zarr-Python 2.* made an attempt to align its API with that of [h5py](https://docs.h5py.org/en/stable/index.html). With 3.0, we will relax this alignment in favor of providing an explicit compatibility module (`zarr.h5py_compat`). This module will expose the `Group` and `Dataset` APIs that map to Zarr-Python’s `Group` and `Array` objects. 
+
+### Creation API
+
+Zarr-Python 2.* bundles together the creation and serialization of Zarr objects. Zarr-Python 3.* will make it possible to create objects in memory separate from serializing them. This will specifically enable writing hierarchies of Zarr objects in a single batch step. For example:
+
+```python
+
+arr1 = Array(shape=(10, 10), path="foo/bar", dtype="i4", store=store)
+arr2 = Array(shape=(10, 10), path="foo/spam", dtype="f8", store=store)
+
+arr1.save()
+arr2.save()
+
+# or equivalently
+
+zarr.save_many([arr1 ,arr2])
+```
+
+### Plugin API
+
+Zarr V3 was designed to be extensible at multiple layers. Zarr-Python will support these extensions through a combination of [Abstract Base Classes](https://docs.python.org/3/library/abc.html) (ABCs) and [Entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/). 
+
+**ABCs**
+
+Zarr V3 will expose Abstract base classes for the following objects:
+
+- `Store`, `ReadStore`, `ReadWriteStore`, `ReadListStore`, and `ReadWriteListStore`
+- `BaseArray`, `SynchronousArray`, and `AsynchronousArray`
+- `BaseGroup`, `SynchronousGroup`, and `AsynchronousGroup`
+- `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`
```

Comment:
@normanrz mentioned the idea about writing a set of codec that would wrap `numcodecs` existing compressors/filters/etc. That would likely be the cleanest way to continue using numcodecs going forward. With the entrypoint api for codecs, the wrapper class could actually live in numcodecs and simply be registered as an entrypoint that follows the protocol. See #1588 for a start on this api.

Reply:
This also came up in today's refactor WG call. @normanrz has given this more thought than I have. Perhaps he can share a bit more on what you think is needed here?

Reply:
The Zarr 3 spec is pretty explicit about what codecs are in spec to ensure compatibility among implementations. That set of codecs is now implemented directly in zarr-python (with use of numcodecs).

However, it is desirable to experiment with other codecs in zarr-python and make use of numcodecs. I think there are 2 ways of doing this:
1. We could make generic adapters for numcodecs to fit the zarr codecs API (ie. ArrayArrayCodec, ArrayBytesCodec, BytesBytesCodec). That would also be useful to code v2 compatibility.
2. We could also introduce a new API in numcodecs that presents all codecs using the Zarr codec API. Not sure what the appetite for such a change on the numcodecs side is.

I am a bit concerned about the codec metadata that gets written to the jsons because most codecs in numcodecs are not in the Zarr v3 spec (or with different metadata). Maybe we should actually mark them as numcodecs codecs:
```
"codecs": [{"name": "numcodecs", "configuration": {"id":"blosc", "cname": "zstd", ...}}]
```

Reply:
Thanks for raising this, @jeromekelleher, and the explanation here, @jhamman and @normanrz.

I was wondering what would be the ideal way to add a new codec (if and when we plan to add any) in Zarr-Python V3.0?
And what would be the best way forward out of the two options if we consider the possibility of extending the existing codecs in the V3 specification?


Also, it'd be great if we could get @jakirkham views on this.

Reply:
To add new codecs to zarr-python, developers can subclass the base classes (e.g. `ArrayBytesCodec`, `ArrayArrayCodec`, `BytesBytesCodec`) and register the codec using the `[zarr.codecs]` entry point in their pyproject.toml or programmatically call [`zarr.v3.codecs.registry.register_codec`](https://github.com/zarr-developers/zarr-python/blob/v3/zarr/v3/codecs/registry.py#L31).